File: backend/app/routers/sessions.py
import json
import os
import uuid
from datetime import datetime, timezone
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File
from sqlalchemy import select, func
from sqlalchemy.orm import selectinload
from sqlalchemy.ext.asyncio import AsyncSession
from app.config import get_settings
from app.db.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.models.session import Session
from app.models.exchange import Exchange
from app.models.integrity import Score, Baseline, IntegrityEvent
from app.schemas.schemas import (
    SessionCreate, SessionResponse, SessionSummary,
    CalibrationCompleteRequest, BaselineResponse
)
from app.services.interviewer_agent import generate_first_question
from app.services.transcription_service import transcribe_audio
from app.services.speech_analyzer import analyze_speech

router = APIRouter(prefix="/sessions", tags=["Sessions"])


@router.post("/create", response_model=SessionResponse, status_code=status.HTTP_201_CREATED)
async def create_session(
    request: SessionCreate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Create a new interview session and generate the first question."""
    session = Session(
        user_id=user.id,
        topic=request.topic,
        difficulty=request.difficulty,
        duration_minutes=request.duration_minutes,
        total_questions=request.total_questions,
    )
    db.add(session)
    await db.flush()

    # Generate the first question
    first_question = await generate_first_question(
        topic=request.topic,
        difficulty=request.difficulty,
    )

    # Create the first exchange (question only, no answer yet)
    exchange = Exchange(
        session_id=session.id,
        question=first_question,
        question_index=1,
    )
    db.add(exchange)
    await db.flush()

    # Eagerly re-load with exchanges + scores
    result = await db.execute(
        select(Session)
        .where(Session.id == session.id)
        .options(selectinload(Session.exchanges).selectinload(Exchange.score))
    )
    session = result.scalar_one()

    return SessionResponse.model_validate(session)


@router.get("/{session_id}", response_model=SessionResponse)
async def get_session(
    session_id: UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get session details with all exchanges."""
    result = await db.execute(
        select(Session)
        .where(Session.id == session_id, Session.user_id == user.id)
        .options(selectinload(Session.exchanges).selectinload(Exchange.score))
    )
    session = result.scalar_one_or_none()

    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session not found",
        )

    return SessionResponse.model_validate(session)


@router.patch("/{session_id}/end", response_model=SessionResponse)
async def end_session(
    session_id: UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """End an active session."""
    result = await db.execute(
        select(Session)
        .where(Session.id == session_id, Session.user_id == user.id)
        .options(selectinload(Session.exchanges).selectinload(Exchange.score))
    )
    session = result.scalar_one_or_none()

    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session not found",
        )

    if session.status != "active":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Session is already ended",
        )

    session.status = "completed"
    session.ended_at = datetime.now(timezone.utc)
    await db.flush()

    return SessionResponse.model_validate(session)


@router.post("/{session_id}/calibration/submit")
async def submit_calibration_audio(
    session_id: UUID,
    audio: UploadFile = File(...),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # Validate session exists and belongs to user
    result = await db.execute(
        select(Session).where(Session.id == session_id, Session.user_id == user.id)
    )
    session = result.scalar_one_or_none()
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    if not audio.filename:
        raise HTTPException(status_code=400, detail="No audio file provided")

    content = await audio.read()
    settings = get_settings()
    calib_dir = os.path.join(settings.UPLOAD_DIR, "calibration", str(session_id))
    os.makedirs(calib_dir, exist_ok=True)
    
    # Generate unique path
    import uuid as uuid_pkg
    audio_path = os.path.join(calib_dir, f"{uuid_pkg.uuid4()}.webm")
    with open(audio_path, "wb") as f:
        f.write(content)

    # Transcribe and analyze
    transcription_result = await transcribe_audio(audio_path)
    speech_analysis = analyze_speech(
        word_timestamps=transcription_result.word_timestamps,
        transcript=transcription_result.transcript,
    )

    return {
        "wpm": speech_analysis.avg_wpm,
        "transcript": transcription_result.transcript,
    }


@router.post("/{session_id}/calibration/complete")
async def complete_calibration(
    session_id: UUID,
    request: CalibrationCompleteRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # Validate session
    result = await db.execute(
        select(Session).where(Session.id == session_id, Session.user_id == user.id)
    )
    session = result.scalar_one_or_none()
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    # Check if baseline already exists
    baseline_result = await db.execute(
        select(Baseline).where(Baseline.session_id == session_id)
    )
    baseline = baseline_result.scalar_one_or_none()

    head_pose_json = json.dumps(request.head_pose_range.model_dump())

    if baseline:
        baseline.avg_wpm = request.avg_wpm
        baseline.wpm_std_dev = request.wpm_std_dev
        baseline.gaze_center_x = request.gaze_center_x
        baseline.gaze_center_y = request.gaze_center_y
        baseline.gaze_std_dev = request.gaze_std_dev
        baseline.head_pose_range = head_pose_json
    else:
        baseline = Baseline(
            session_id=session_id,
            avg_wpm=request.avg_wpm,
            wpm_std_dev=request.wpm_std_dev,
            gaze_center_x=request.gaze_center_x,
            gaze_center_y=request.gaze_center_y,
            gaze_std_dev=request.gaze_std_dev,
            head_pose_range=head_pose_json,
        )
        db.add(baseline)

    await db.flush()
    return {"status": "success", "message": "Calibration baseline saved successfully"}


@router.get("/{session_id}/baseline", response_model=BaselineResponse)
async def get_session_baseline(
    session_id: UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # Validate session
    result = await db.execute(
        select(Session).where(Session.id == session_id, Session.user_id == user.id)
    )
    session = result.scalar_one_or_none()
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    baseline_result = await db.execute(
        select(Baseline).where(Baseline.session_id == session_id)
    )
    baseline = baseline_result.scalar_one_or_none()
    if not baseline:
        raise HTTPException(status_code=404, detail="Baseline calibration not found for this session")

    return baseline


@router.get("/{session_id}/report")
async def get_session_report(
    session_id: UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # Fetch session with exchanges and scores
    result = await db.execute(
        select(Session)
        .where(Session.id == session_id, Session.user_id == user.id)
        .options(selectinload(Session.exchanges).selectinload(Exchange.score))
    )
    session = result.scalar_one_or_none()
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    # Fetch all integrity events
    events_result = await db.execute(
        select(IntegrityEvent)
        .where(IntegrityEvent.session_id == session_id)
        .order_by(IntegrityEvent.timestamp_ms)
    )
    events = events_result.scalars().all()

    # Determine Verdict
    verdict = "EXCELLENT"
    if session.integrity_score < 70:
        verdict = "SUSPICIOUS"
    elif session.integrity_score < 85:
        verdict = "REVIEW_RECOMMENDED"

    # Construct timeline
    timeline = []
    session_start_ms = int(session.started_at.replace(tzinfo=timezone.utc).timestamp() * 1000)

    # Group events by exchange_id (or general if none)
    exchange_events = {}
    for evt in events:
        eid = evt.exchange_id or "general"
        if eid not in exchange_events:
            exchange_events[eid] = []
        
        # Format relative timestamp MM:SS
        rel_ms = evt.timestamp_ms - session_start_ms
        rel_sec = max(0, int(rel_ms // 1000))
        time_str = f"{rel_sec // 60:02d}:{rel_sec % 60:02d}"

        try:
            meta = json.loads(evt.metadata_json) if evt.metadata_json else {}
        except Exception:
            meta = {}

        exchange_events[eid].append({
            "type": evt.event_type,
            "severity": evt.severity,
            "timestamp": time_str,
            "metadata": meta
        })

    # For each exchange in session, build question record with its flags
    for ex in session.exchanges:
        flags = exchange_events.get(ex.id, [])
        suspicion = "low"
        if any(f["severity"] == "high" for f in flags):
            suspicion = "high"
        elif any(f["severity"] == "medium" for f in flags) or len(flags) > 2:
            suspicion = "moderate"

        timeline.append({
            "question_index": ex.question_index,
            "question": ex.question,
            "flags": flags,
            "suspicion_level": suspicion
        })

    # Add general/calibration phase flags if any
    general_flags = exchange_events.get("general", [])
    if general_flags:
        timeline.insert(0, {
            "question_index": 0,
            "question": "Baseline Calibration & Setup",
            "flags": general_flags,
            "suspicion_level": "moderate" if any(f["severity"] == "high" or f["severity"] == "medium" for f in general_flags) else "low"
        })

    # Gather correlations
    correlations = [
        ex.score.gaze_fluency_correlation 
        for ex in session.exchanges 
        if ex.score and ex.score.gaze_fluency_correlation is not None
    ]
    avg_correlation = sum(correlations) / len(correlations) if correlations else 0.0

    # Count consistent deviation directions
    deviation_types = [evt.event_type for evt in events if evt.event_type in ["notes_below_camera", "second_monitor_left", "second_monitor_right"]]
    dominant_suspicious_zone = "none"
    if deviation_types:
        dominant_suspicious_zone = max(set(deviation_types), key=deviation_types.count)

    summary_text = "No significant cheating patterns detected."
    if session.integrity_score < 100:
        if avg_correlation > 0.5:
            summary_text = f"Candidate showed a notable gaze-fluency correlation ({avg_correlation:.2f}) preceding fluent answers. "
        else:
            summary_text = "Candidate showed behavioral anomalies. "
        
        if dominant_suspicious_zone != "none":
            friendly_zone = dominant_suspicious_zone.replace("_", " ")
            summary_text += f"Dominant suspicious zone: {friendly_zone}."

    return {
        "session_id": session.id,
        "integrity_score": session.integrity_score,
        "verdict": verdict,
        "timeline": timeline,
        "pattern_analysis": {
            "gaze_fluency_correlation": round(avg_correlation, 2),
            "consistent_deviation_direction": len(set(deviation_types)) == 1 if deviation_types else False,
            "dominant_suspicious_zone": dominant_suspicious_zone,
            "summary": summary_text
        }
    }

File: backend/src/routes/sessions.ts
import { Router, Response } from 'express';
import { v4 as uuidv4 } from 'uuid';
import path from 'path';
import fs from 'fs';
import multer from 'multer';
import { getDb } from '../db';
import { requireAuth, AuthRequest } from '../middleware/auth';
import { generateFirstQuestion } from '../services/interviewerAgent';
import { transcribeAudio } from '../services/transcriptionService';
import { analyzeSpeech } from '../services/speechAnalyzer';
import { config } from '../config';

const router = Router();

// Setup multer for temporary file uploads
const upload = multer({
  dest: path.resolve(config.UPLOAD_DIR, 'temp'),
  limits: { fileSize: config.MAX_AUDIO_SIZE_MB * 1024 * 1024 }
});

// POST /sessions/create
router.post('/create', requireAuth, async (req: AuthRequest, res: Response) => {
  const { topic, difficulty, duration_minutes, total_questions, llm_provider, llm_model } = req.body;

  if (!topic) {
    return res.status(400).json({ detail: 'Topic is required' });
  }

  try {
    const db = getDb();
    const sessionId = uuidv4();
    const startedAt = new Date().toISOString();

    const provider = llm_provider || config.DEFAULT_LLM_PROVIDER;
    const model = llm_model || config.DEFAULT_LLM_MODEL;

    // Create session record
    await db.run(
      `INSERT INTO sessions 
        (id, user_id, topic, difficulty, duration_minutes, status, integrity_score, total_questions, started_at, llm_provider, llm_model) 
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        sessionId,
        req.userId,
        topic,
        difficulty || 'medium',
        duration_minutes || 30,
        'active',
        100,
        total_questions || 10,
        startedAt,
        provider,
        model
      ]
    );

    // Generate first question
    const firstQuestion = await generateFirstQuestion(topic, difficulty || 'medium', provider, model, req.user);

    // Create first exchange record (index 1)
    const exchangeId = uuidv4();
    await db.run(
      'INSERT INTO exchanges (id, session_id, question, question_index, created_at) VALUES (?, ?, ?, ?, ?)',
      [exchangeId, sessionId, firstQuestion, 1, startedAt]
    );

    // Load session with exchanges to return
    const session = await db.get('SELECT * FROM sessions WHERE id = ?', [sessionId]);
    const exchanges = await db.all('SELECT * FROM exchanges WHERE session_id = ? ORDER BY question_index ASC', [sessionId]);

    return res.status(201).json({
      ...session,
      exchanges
    });
  } catch (error) {
    console.error('Session creation error:', error);
    return res.status(500).json({ detail: 'Internal server error during session creation' });
  }
});

// GET /sessions/:id
router.get('/:id', requireAuth, async (req: AuthRequest, res: Response) => {
  const { id } = req.params;

  try {
    const db = getDb();
    const session = await db.get('SELECT * FROM sessions WHERE id = ? AND user_id = ?', [id, req.userId]);

    if (!session) {
      return res.status(404).json({ detail: 'Session not found' });
    }

    // Load exchanges and their corresponding scores
    const exchanges = await db.all(
      `SELECT e.*, 
              s.technical_accuracy, s.definition_present, s.mechanism_explained, s.example_given, 
              s.edge_cases_mentioned, s.missing_concepts, s.follow_up_angle, s.wpm, s.filler_count, 
              s.longest_pause_seconds, s.confidence_proxy, s.gaze_fluency_correlation
       FROM exchanges e
       LEFT JOIN scores s ON s.exchange_id = e.id
       WHERE e.session_id = ?
       ORDER BY e.question_index ASC`,
      [id]
    );

    const formattedExchanges = exchanges.map(ex => {
      const score = ex.technical_accuracy !== null ? {
        id: ex.id,
        exchange_id: ex.id,
        technical_accuracy: ex.technical_accuracy,
        definition_present: !!ex.definition_present,
        mechanism_explained: !!ex.mechanism_explained,
        example_given: !!ex.example_given,
        edge_cases_mentioned: !!ex.edge_cases_mentioned,
        missing_concepts: ex.missing_concepts ? JSON.parse(ex.missing_concepts) : null,
        follow_up_angle: ex.follow_up_angle,
        wpm: ex.wpm,
        filler_count: ex.filler_count,
        longest_pause_seconds: ex.longest_pause_seconds,
        confidence_proxy: ex.confidence_proxy,
        gaze_fluency_correlation: ex.gaze_fluency_correlation
      } : null;

      return {
        id: ex.id,
        session_id: ex.session_id,
        question: ex.question,
        answer_transcript: ex.answer_transcript,
        question_index: ex.question_index,
        created_at: ex.created_at,
        score
      };
    });

    return res.json({
      ...session,
      exchanges: formattedExchanges
    });
  } catch (error) {
    console.error('Error fetching session:', error);
    return res.status(500).json({ detail: 'Internal server error fetching session details' });
  }
});

// PATCH /sessions/:id/end
router.patch('/:id/end', requireAuth, async (req: AuthRequest, res: Response) => {
  const { id } = req.params;

  try {
    const db = getDb();
    const session = await db.get('SELECT * FROM sessions WHERE id = ? AND user_id = ?', [id, req.userId]);

    if (!session) {
      return res.status(404).json({ detail: 'Session not found' });
    }

    if (session.status !== 'active') {
      return res.status(400).json({ detail: 'Session is already ended' });
    }

    const endedAt = new Date().toISOString();
    await db.run('UPDATE sessions SET status = ?, ended_at = ? WHERE id = ?', ['completed', endedAt, id]);

    const updatedSession = await db.get('SELECT * FROM sessions WHERE id = ?', [id]);
    const exchanges = await db.all('SELECT * FROM exchanges WHERE session_id = ? ORDER BY question_index ASC', [id]);

    return res.json({
      ...updatedSession,
      exchanges
    });
  } catch (error) {
    console.error('Error ending session:', error);
    return res.status(500).json({ detail: 'Internal server error while ending session' });
  }
});

// POST /sessions/:id/calibration/submit
router.post('/:id/calibration/submit', requireAuth, upload.single('audio'), async (req: AuthRequest, res: Response) => {
  const { id } = req.params;

  if (!req.file) {
    return res.status(400).json({ detail: 'No audio file provided' });
  }

  try {
    const db = getDb();
    const session = await db.get('SELECT * FROM sessions WHERE id = ? AND user_id = ?', [id, req.userId]);

    if (!session) {
      fs.unlinkSync(req.file.path);
      return res.status(404).json({ detail: 'Session not found' });
    }

    // Save audio permanently
    const calibDir = path.join(config.UPLOAD_DIR, 'calibration', id);
    fs.mkdirSync(calibDir, { recursive: true });

    const permanentPath = path.join(calibDir, `${uuidv4()}.webm`);
    fs.renameSync(req.file.path, permanentPath);

    // Transcribe and analyze speech
    const transcriptionResult = await transcribeAudio(permanentPath, req.user, 2, session.llm_provider);
    const speechAnalysis = analyzeSpeech(
      transcriptionResult.word_timestamps,
      transcriptionResult.transcript
    );

    return res.json({
      wpm: speechAnalysis.avg_wpm,
      transcript: transcriptionResult.transcript,
    });
  } catch (error) {
    console.error('Error submitting calibration:', error);
    if (req.file && fs.existsSync(req.file.path)) {
      fs.unlinkSync(req.file.path);
    }
    return res.status(500).json({ detail: 'Internal server error processing calibration audio' });
  }
});

// POST /sessions/:id/calibration/complete
router.post('/:id/calibration/complete', requireAuth, async (req: AuthRequest, res: Response) => {
  const { id } = req.params;
  const { avg_wpm, wpm_std_dev, gaze_center_x, gaze_center_y, gaze_std_dev, head_pose_range } = req.body;

  try {
    const db = getDb();
    const session = await db.get('SELECT * FROM sessions WHERE id = ? AND user_id = ?', [id, req.userId]);

    if (!session) {
      return res.status(404).json({ detail: 'Session not found' });
    }

    const headPoseJson = JSON.stringify(head_pose_range);

    const existingBaseline = await db.get('SELECT id FROM baselines WHERE session_id = ?', [id]);
    if (existingBaseline) {
      await db.run(
        `UPDATE baselines 
         SET avg_wpm = ?, wpm_std_dev = ?, gaze_center_x = ?, gaze_center_y = ?, gaze_std_dev = ?, head_pose_range = ? 
         WHERE session_id = ?`,
        [avg_wpm, wpm_std_dev, gaze_center_x, gaze_center_y, gaze_std_dev, headPoseJson, id]
      );
    } else {
      const baselineId = uuidv4();
      const capturedAt = new Date().toISOString();
      await db.run(
        `INSERT INTO baselines 
          (id, session_id, avg_wpm, wpm_std_dev, gaze_center_x, gaze_center_y, gaze_std_dev, head_pose_range, captured_at) 
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        [baselineId, id, avg_wpm, wpm_std_dev, gaze_center_x, gaze_center_y, gaze_std_dev, headPoseJson, capturedAt]
      );
    }

    return res.json({ status: 'success', message: 'Calibration baseline saved successfully' });
  } catch (error) {
    console.error('Error saving calibration baseline:', error);
    return res.status(500).json({ detail: 'Internal server error saving calibration details' });
  }
});

// GET /sessions/:id/baseline
router.get('/:id/baseline', requireAuth, async (req: AuthRequest, res: Response) => {
  const { id } = req.params;

  try {
    const db = getDb();
    const session = await db.get('SELECT * FROM sessions WHERE id = ? AND user_id = ?', [id, req.userId]);

    if (!session) {
      return res.status(404).json({ detail: 'Session not found' });
    }

    const baseline = await db.get('SELECT * FROM baselines WHERE session_id = ?', [id]);
    if (!baseline) {
      return res.status(404).json({ detail: 'Baseline calibration not found for this session' });
    }

    // Parse head_pose_range back
    if (baseline.head_pose_range) {
      baseline.head_pose_range = JSON.parse(baseline.head_pose_range);
    }

    return res.json(baseline);
  } catch (error) {
    console.error('Error fetching baseline:', error);
    return res.status(500).json({ detail: 'Internal server error fetching baseline' });
  }
});

// GET /sessions/:id/report
router.get('/:id/report', requireAuth, async (req: AuthRequest, res: Response) => {
  const { id } = req.params;

  try {
    const db = getDb();
    const session = await db.get('SELECT * FROM sessions WHERE id = ? AND user_id = ?', [id, req.userId]);

    if (!session) {
      return res.status(404).json({ detail: 'Session not found' });
    }

    const exchanges = await db.all(
      `SELECT e.*, s.gaze_fluency_correlation, s.technical_accuracy
       FROM exchanges e
       LEFT JOIN scores s ON s.exchange_id = e.id
       WHERE e.session_id = ?
       ORDER BY e.question_index ASC`,
      [id]
    );

    const events = await db.all(
      'SELECT * FROM integrity_events WHERE session_id = ? ORDER BY timestamp_ms ASC',
      [id]
    );

    // Determine Verdict
    let verdict = 'EXCELLENT';
    const score = session.integrity_score;
    if (score < 70) {
      verdict = 'SUSPICIOUS';
    } else if (score < 85) {
      verdict = 'REVIEW_RECOMMENDED';
    }

    // Build timeline
    const sessionStartMs = new Date(session.started_at + ' Z').getTime();
    const exchangeEventsMap: Record<string, any[]> = {};

    for (const evt of events) {
      const key = evt.exchange_id || 'general';
      if (!exchangeEventsMap[key]) {
        exchangeEventsMap[key] = [];
      }

      const relMs = Number(evt.timestamp_ms) - sessionStartMs;
      const relSec = Math.max(0, Math.floor(relMs / 1000));
      const mm = String(Math.floor(relSec / 60)).padStart(2, '0');
      const ss = String(relSec % 60).padStart(2, '0');
      const timeStr = `${mm}:${ss}`;

      let metadata = {};
      try {
        metadata = evt.metadata_json ? JSON.parse(evt.metadata_json) : {};
      } catch (err) {}

      exchangeEventsMap[key].push({
        type: evt.event_type,
        severity: evt.severity,
        timestamp: timeStr,
        metadata,
      });
    }

    const timeline: any[] = [];

    // Add general/calibration phase flags if any
    const generalFlags = exchangeEventsMap['general'] || [];
    if (generalFlags.length > 0) {
      const hasSevere = generalFlags.some(f => f.severity === 'high' || f.severity === 'medium');
      timeline.push({
        question_index: 0,
        question: 'Baseline Calibration & Setup',
        flags: generalFlags,
        suspicion_level: hasSevere ? 'moderate' : 'low',
      });
    }

    for (const ex of exchanges) {
      const flags = exchangeEventsMap[ex.id] || [];
      let suspicion = 'low';
      if (flags.some(f => f.severity === 'high')) {
        suspicion = 'high';
      } else if (flags.some(f => f.severity === 'medium') || flags.length > 2) {
        suspicion = 'moderate';
      }

      timeline.push({
        question_index: ex.question_index,
        question: ex.question,
        flags,
        suspicion_level: suspicion,
      });
    }

    // Gather correlations
    const correlations = exchanges
      .map(e => e.gaze_fluency_correlation)
      .filter((c): c is number => c !== null && c !== undefined);

    const avgCorrelation =
      correlations.length > 0 ? correlations.reduce((a, b) => a + b, 0) / correlations.length : 0.0;

    // Count consistent deviation directions
    const deviationTypes = events
      .map(e => e.event_type)
      .filter(t => ['notes_below_camera', 'second_monitor_left', 'second_monitor_right'].includes(t));

    let dominantSuspiciousZone = 'none';
    if (deviationTypes.length > 0) {
      // Find mode
      const frequencies: Record<string, number> = {};
      let maxFreq = 0;
      for (const t of deviationTypes) {
        frequencies[t] = (frequencies[t] || 0) + 1;
        if (frequencies[t] > maxFreq) {
          maxFreq = frequencies[t];
          dominantSuspiciousZone = t;
        }
      }
    }

    let summaryText = 'No significant cheating patterns detected.';
    if (score < 100) {
      if (avgCorrelation > 0.5) {
        summaryText = `Candidate showed a notable gaze-fluency correlation (${avgCorrelation.toFixed(2)}) preceding fluent answers. `;
      } else {
        summaryText = 'Candidate showed behavioral anomalies. ';
      }

      if (dominantSuspiciousZone !== 'none') {
        const friendlyZone = dominantSuspiciousZone.replace(/_/g, ' ');
        summaryText += `Dominant suspicious zone: ${friendlyZone}.`;
      }
    }

    return res.json({
      session_id: session.id,
      integrity_score: score,
      verdict,
      timeline,
      pattern_analysis: {
        gaze_fluency_correlation: Math.round(avgCorrelation * 100) / 100,
        consistent_deviation_direction: new Set(deviationTypes).size === 1 && deviationTypes.length > 0,
        dominant_suspicious_zone: dominantSuspiciousZone,
        summary: summaryText,
      },
    });
  } catch (error) {
    console.error('Error generating report:', error);
    return res.status(500).json({ detail: 'Internal server error generating report' });
  }
});

export default router;
